File: ios/Sources/Mia21/Models/ConfigurationModels.swift
//
//  ConfigurationModels.swift
//  Mia21
//
//  Created on November 4, 2025.
//  Copyright © 2025 Mia21. All rights reserved.
//
//  Description:
//  Configuration and initialization models.
//  Includes response modes, initialization options, voice configuration,
//  and initialization responses.
//

import Foundation

// MARK: - Environment

public enum Mia21Environment {
  case production
  case staging
  case custom(String)
  
  public var baseURL: String {
    switch self {
    case .production:
      return "https://api.mia21.com"
    case .staging:
      return "https://api-staging.mia21.com"
    case .custom(let url):
      return url
    }
  }
}

// MARK: - Response Mode

public enum ResponseMode: String, Codable {
  case text = "text"
  case streamText = "stream_text"
  case streamVoice = "stream_voice"
  case streamVoiceOnly = "stream_voice_only"
}

// MARK: - Initialize Options

public struct InitializeOptions {
  public var spaceId: String?
  public var botId: String?
  public var llmType: LLMType?
  public var userName: String?
  public var language: String?
  public var generateFirstMessage: Bool
  public var incognitoMode: Bool
  public var customerLlmKey: String?
  public var spaceConfig: SpaceConfig?

  public init(
    spaceId: String? = nil,
    botId: String? = nil,
    llmType: LLMType? = .openai,
    userName: String? = nil,
    language: String? = nil,
    generateFirstMessage: Bool = true,
    incognitoMode: Bool = false,
    customerLlmKey: String? = nil,
    spaceConfig: SpaceConfig? = nil
  ) {
    self.spaceId = spaceId
    self.botId = botId
    self.llmType = llmType
    self.userName = userName
    self.language = language
    self.generateFirstMessage = generateFirstMessage
    self.incognitoMode = incognitoMode
    self.customerLlmKey = customerLlmKey
    self.spaceConfig = spaceConfig
  }
}

// MARK: - Initialize Response

public struct InitializeResponse: Codable {
  public let status: String
  public let userId: String
  public let message: String?
  public let spaceId: String?
  public let isNewUser: Bool?
  
  // Note: No custom CodingKeys needed - APIClient uses .convertFromSnakeCase
}

// MARK: - Voice Configuration

public struct VoiceConfig: Codable {
  public let enabled: Bool
  public let voiceId: String?
  public let elevenlabsApiKey: String?
  public let stability: Double?
  public let similarityBoost: Double?
  
  // Note: No custom CodingKeys needed - APIClient uses .convertFromSnakeCase

  public init(
    enabled: Bool,
    voiceId: String? = nil,
    elevenlabsApiKey: String? = nil,
    stability: Double? = nil,
    similarityBoost: Double? = nil
  ) {
    self.enabled = enabled
    self.voiceId = voiceId
    self.elevenlabsApiKey = elevenlabsApiKey
    self.stability = stability
    self.similarityBoost = similarityBoost
  }
}

File: android/lib/src/main/kotlin/com/mia21/models/ConfigurationModels.kt
/**
 * ConfigurationModels.kt
 * Mia21
 *
 * Created on November 25, 2025.
 * Copyright © 2025 Mia21. All rights reserved.
 *
 * Description:
 * Configuration and initialization models.
 * Includes response modes, initialization options, voice configuration,
 * and initialization responses.
 */

package com.mia21.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * API environment configuration
 */
enum class Mia21Environment(val baseURL: String, val environmentName: String) {
    PRODUCTION("https://api.mia21.com", "Production"),
    STAGING("https://api-staging.mia21.com", "Staging");
}

/**
 * Response mode for API requests
 */
@Serializable
enum class ResponseMode {
    @SerialName("text")
    TEXT,
    
    @SerialName("stream_text")
    STREAM_TEXT,
    
    @SerialName("stream_voice")
    STREAM_VOICE,
    
    @SerialName("stream_voice_only")
    STREAM_VOICE_ONLY
}

/**
 * Options for initializing a chat session
 */
data class InitializeOptions(
    val spaceId: String? = null,
    val botId: String? = null,
    val llmType: LLMType? = LLMType.OPENAI,
    val userName: String? = null,
    val language: String? = null,
    val generateFirstMessage: Boolean = true,
    val incognitoMode: Boolean = false,
    val customerLlmKey: String? = null,
    val spaceConfig: SpaceConfig? = null
)

/**
 * Response from chat initialization
 */
@Serializable
data class InitializeResponse(
    val status: String,
    @SerialName("user_id")
    val userId: String,
    @SerialName("conversation_id")
    val conversationId: String,
    val message: String? = null,
    @SerialName("space_id")
    val spaceId: String? = null,
    @SerialName("is_new_user")
    val isNewUser: Boolean? = null
)

/**
 * Voice synthesis configuration
 */
@Serializable
data class VoiceConfig(
    val enabled: Boolean,
    @SerialName("voice_id")
    val voiceId: String? = null,
    @SerialName("elevenlabs_api_key")
    val elevenlabsApiKey: String? = null,
    val stability: Double? = null,
    @SerialName("similarity_boost")
    val similarityBoost: Double? = null
)

